src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutInfoNode.cs

```diff
@@ -0,0 +1,305 @@
+
+        private NativeWriter _writer;
+        private MemoryStream _writerStream;
+
+        private Section _signaturesSection;
+        private Section _ldTokenInfoSection;
+
+        public NativeLayoutInfoNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__nativelayoutinfo_End", true);
+            _externalReferences = externalReferences;
+            _writer = new NativeWriter();
+            _signaturesSection = _writer.NewSection();
+            _ldTokenInfoSection = _writer.NewSection();
+        }
+
+        public ISymbolNode EndSymbol
+        {
+            get
+            {
+                return _endSymbol;
+            }
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__nativelayoutinfo");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.DataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName() => this.GetMangledName();
+
+        public void SaveNativeLayoutInfoWriter()
+        {
+            if (_writerStream != null)
+            {
+#if DEBUG
+                // Sanity check... We should not write new items to the native layout after 
+                // we've already saved it.
+
+                MemoryStream debugStream = new MemoryStream();
+                _writer.Save(debugStream);
+                byte[] debugStreamBytes = debugStream.ToArray();
+                byte[] nativeLayoutInfoBytes = _writerStream.ToArray();
+                Debug.Assert(debugStreamBytes.Length == nativeLayoutInfoBytes.Length);
+                for (int i = 0; i < debugStreamBytes.Length; i++)
+                    Debug.Assert(debugStreamBytes[i] == nativeLayoutInfoBytes[i]);
+#endif
+                return;
+            }
+
+            _writerStream = new MemoryStream();
+            _writer.Save(_writerStream);
+        }
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+
+            SaveNativeLayoutInfoWriter();
+
+            byte[] nativeLayoutInfoBytes = _writerStream.ToArray();
+
+            _endSymbol.SetSymbolOffset(nativeLayoutInfoBytes.Length);
+
+            return new ObjectData(nativeLayoutInfoBytes, Array.Empty<Relocation>(), 1, new ISymbolNode[] { this, _endSymbol });
+        }
+
+        #region Vertex Building Functions
+        public Vertex GetNativeLayoutInfoSignatureForLdToken<T>(NodeFactory factory, T methodDescOrFieldDesc)
```

Comment:
There does not seem to be any benefits of this being generic method. Should it rather be two methods - one for MethodDesc and one for FieldDesc?